From 9f282784d377eb7e778a99c9014337c8f14d35b5 Mon Sep 17 00:00:00 2001
From: Quanyang Wang <quanyang.wang@windriver.com>
Date: Wed, 22 May 2019 17:13:07 +0800
Subject: [PATCH 3/3] soc: imx: gpc-psci: fix kdump hang issue by using
 cpu_logical_map(cpu)

In imx8mq, the GPC(General Power Controller) controls power managerment
of the chip, including mask/unmask interrupt wakeup source for cpu0-3.

In kdump, the dump-capture kernel may not run at cpu0, it means that
the logical cpu0 maybe the hardware cpu1-3. Sometimes dump-capture kernel
run at hwcpu 1 but all irqs are assigned to be the wakeup source of cpu0.
It result that when call "usleep" in dump-capture kernel, the cpu1 will
never wakeup because there is no irq to disturb it.

So we should set all irq to be the wakeup source of hwcpu instead of
logical cpu.

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/soc/imx/gpc-psci.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/soc/imx/gpc-psci.c b/drivers/soc/imx/gpc-psci.c
index d3867e18ee23..667801aab973 100644
--- a/drivers/soc/imx/gpc-psci.c
+++ b/drivers/soc/imx/gpc-psci.c
@@ -28,6 +28,7 @@
 #include <linux/platform_device.h>
 #include <linux/regulator/consumer.h>
 #include <linux/pm_domain.h>
+#include <asm/smp_plat.h>
 #include <soc/imx/fsl_sip.h>
 
 #define GPC_MAX_IRQS		(4 * 32)
@@ -98,7 +99,7 @@ static int imx_gpc_psci_irq_set_affinity(struct irq_data *d,
 
 	spin_lock(&gpc_psci_lock);
 	arm_smccc_smc(FSL_SIP_GPC, 0x4, d->hwirq,
-		      cpu, 0, 0, 0, 0, &res);
+		      cpu_logical_map(cpu), 0, 0, 0, 0, &res);
 	spin_unlock(&gpc_psci_lock);
 
 	return 0;
-- 
2.17.1

